frappe/types/filter.py

```diff
@@ -0,0 +1,280 @@
+import textwrap
+from collections import defaultdict
+from collections.abc import Generator, Iterable, Mapping, Sequence
+from datetime import date, datetime
+from itertools import groupby
+from operator import attrgetter
+from typing import Any, NamedTuple, TypeAlias, TypeGuard, TypeVar, cast
+
+from pypika import Column
+from typing_extensions import Self, override
+
+from .docref import DocRef
+
+Doct: TypeAlias = str
+Fld: TypeAlias = str
+Op: TypeAlias = str
+DateTime: TypeAlias = datetime | date
+_Val: TypeAlias = str | int | None | DateTime | Column
+_InVal: TypeAlias = _Val | DocRef | bool
+Val: TypeAlias = _Val | Sequence[_Val]
+InVal: TypeAlias = _InVal | Sequence[_InVal]
+
+
+FilterTupleSpec: TypeAlias = tuple[Fld, InVal] | tuple[Fld, Op, InVal] | tuple[Doct, Fld, Op, InVal]
+FilterMappingSpec: TypeAlias = Mapping[Fld, _InVal | tuple[Op, InVal]]
+
+
+class Sentinel:
+	def __bool__(self) -> bool:
+		return False
+
+	@override
+	def __str__(self) -> str:
+		return "UNSPECIFIED"
+
+
+UNSPECIFIED = Sentinel()
+
+T = TypeVar("T")
+
+
+def is_not_unspecified(value: T | Sentinel) -> TypeGuard[T]:
```

Comment:
Check with type checker if logic can be inverted without defeating the TypeGuard